optuna/visualization/matplotlib/_rank.py

```diff
@@ -0,0 +1,156 @@
+def plot_rank(
+    study: Study,
+    params: list[str] | None = None,
+    *,
+    target: Callable[[FrozenTrial], float] | None = None,
+    target_name: str = "Objective Value",
+) -> "Axes":
+    """Plot parameter relations as scatter plots with colors indicating ranks of objective value.
+
+    Note that, if a parameter contains missing values, a trial with missing values is not plotted.
+
+    .. seealso::
+        Please refer to :func:`optuna.visualization.plot_rank` for an example.
+
+    Warnings:
+        Output figures of this Matplotlib-based
+        :func:`~optuna.visualization.matplotlib.plot_rank` function would be different from
+        those of the Plotly-based :func:`~optuna.visualization.plot_rank`.
+
+    Example:
+
+        The following code snippet shows how to plot the parameter relationship as a rank plot.
+
+        .. plot::
+
+            import optuna
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -100, 100)
+                y = trial.suggest_categorical("y", [-1, 0, 1])
+                return x ** 2 + y
+
+
+            sampler = optuna.samplers.TPESampler(seed=10)
+            study = optuna.create_study(sampler=sampler)
+            study.optimize(objective, n_trials=30)
+
+            optuna.visualization.matplotlib.plot_rank(study, params=["x", "y"])
+
+    Args:
+        study:
+            A :class:`~optuna.study.Study` object whose trials are plotted for their target values.
+        params:
+            Parameter list to visualize. The default is all parameters.
+        target:
+            A function to specify the value to display. If it is :obj:`None` and ``study`` is being
+            used for single-objective optimization, the objective values are plotted.
+
+            .. note::
+                Specify this argument if ``study`` is being used for multi-objective optimization.
+        target_name:
+            Target's name to display on the color bar.
+
+    Returns:
+        A :class:`matplotlib.axes.Axes` object.
+    """
+
+    _imports.check()
+    _logger.warning(
+        "Output figures of this Matplotlib-based `plot_rank` function would be different from "
+        "those of the Plotly-based `plot_rank`."
+    )
+    info = _get_rank_info(study, params, target, target_name)
+    return _get_rank_plot(info)
+
+
+def _get_rank_plot(
+    info: _RankPlotInfo,
+) -> "Axes":
+    params = info.params
+    sub_plot_infos = info.sub_plot_infos
+
+    plt.style.use("ggplot")  # Use ggplot style sheet for similar outputs to plotly.
+
+    title = f"Rank ({info.target_name})"
+
+    n_params = len(params)
+    if n_params == 0:
+        _, ax = plt.subplots()
+        ax.set_title(title)
+        return ax
+    if n_params == 1 or n_params == 2:
+        fig, axs = plt.subplots()
+        axs.set_title(title)
+        pc = _add_rank_subplot(axs, sub_plot_infos[0][0])
+    else:
+        fig, axs = plt.subplots(n_params, n_params)
```

Comment:
Nit: Maybe we don't need these `if`s and can directly have `plt.subplots(len(sub_plot_infos), len(sub_plot_infos))`.

Reply:
I think these `if`s are required by the following reason.
* `if n_params == 0:` is required because in this case, `sub_plot_infos` is `[]` and `sub_plot_infos[0][0]` raises error if we remove this `if`.
* `if n_params == 1 or n_params == 2:` is required because `plt.subplots()` or `plt.subplots(1, 1)` returns `Axes` object. On the other hand, `plt.subplots(n, n)` returns array of `Axes` object if `n >= 2`. Thus, `ax = axs[x_i, y_i]` raises error if we remove this `if`.